Extract get event call from PullDhisApiDataSource, and refactor checkAllQuarantineSurveys to use domain entities.

idelcano opened the pull request:
Closes #2016

idelcano committed:
working

idelcano committed:
Merge branch 'development' into feature-create_remote_data_source_to_call_getevents

idelcano committed:
update surveydb queries

idelcano committed:
Update orgunitdb queries

idelcano committed:
Added org unit data source repository and mapper

idelcano committed:
update survey filter

idelcano committed:
modify ISurveyDataSource and datasources

idelcano committed:
modify fix quarantine survey status use case and survey quarantine controller

idelcano committed:
Avoid create new surveys db when mapping

idelcano committed:
use new use case and remove old methods to check quarantine

idelcano committed:
refactor filter

idelcano committed:
refactor UpdateQuarantineSurveyStatusUseCase

idelcano committed:
update and add tests

idelcano committed:
Added date validation

idelcano committed:
modify filter

idelcano committed:
fix test

idelcano committed:
Added flag to control the Use case is only running one time

idelcano commented:
### :pushpin: References
* **Issue:** connect https://github.com/EyeSeeTea/malariapp/issues/2016

### :tophat: What is the goal?

Extract get event call from PullDhisApiDataSource, and refactor checkAllQuarantineSurveys to use domain entities.

### :memo: How is it being implemented?

I refactored the checkQuarantineMethod to use a UseCase.

I added a UpdateQuarantineSurveyStatusUseCase. 

This UseCase get the OrgUnit and Program Uids using a OrgUnitRepository, 
and create the necessary surveyFilters to check the surveys in quarantine using the repository.
The UseCase compares the local surveys against the pulled surveys to set as sent or as completed the quarantine surveys, and save them using the repository.

I added some creation methods and test it in surveyFilter and Survey.

I added a SurveyApiDatasource to get events from api data source, and remove all the old methods.

I have removed some SurveyDB queries and use entities to find the correct parameters to the filters.

I have add surveyQuarantineRepository, OrgUnitRepository and orgUnitDBMapper

I added a flag to avoid multiple Quarantine Use Cases instances because debugging i found this problem and it might make duplicates.


### :boom: How can it be tested?

- [ ] **Use Case 1:**  force some quarantine surveys and check them in the server after checkquarantine and next pushcontroller  was executed

### :floppy_disk: Requires DB migration?

- [x] Nope, we can just merge this branch.
- [ ] Yes, but we need to apply it before merging this branch.
- [ ] Yes, it's already applied.

### :art: UI changes?

- [x] Nope, the UI remains as beautiful as it was before!
- [ ] Yeap, here you have some screenshots-

idelcano committed:
Merge branch 'development' into feature-create_remote_data_source_to_call_getevents

idelcano committed:
refactor some private methods and extract some code

xurxodev reviewed (changes requested):
@idelcano please resolve conflicts and refactor. Review the development branch because has architecture changes.

idelcano committed:
Merge branch 'development' into feature-create_remote_data_source_to_call_getevents

idelcano committed:
fix missing code after merge

idelcano committed:
fix tests

idelcano committed:
avoid unnecesary server queries

idelcano committed:
Merge branch 'development' into feature-create_remote_data_source_to_call_getevents

xurxodev committed:
Merge branch 'development' into feature-create_remote_data_source_to_call_getevents

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/data/database/datasources/OrgUnitLocalDataSource.java:
```diff
@@ -0,0 +1,85 @@
+import static org.eyeseetea.malariacare.data.database.AppDatabase.programAlias;
+import static org.eyeseetea.malariacare.data.database.AppDatabase.programName;
+
+public class OrgUnitLocalDataSource implements IOrgUnitRepository {
```
There are several problems here:
This is a Datasource but implement IOrgUnitRepository
This DataSource is similar to OrgUnitRepository that implements IOrgUnitRepository also
Please refactor to we have Repository, localdatasource according to Clean Architecture with Read Policy

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/data/database/datasources/SurveyLocalDataSource.java:
```diff
@@ -204,4 +205,32 @@ private void loadValuesInSurveys(List<SurveyDB> surveyDBS) {
         }
     }
+
+    public List<Survey> mapQuarantineSurveys(List<SurveyDB> surveysDB) {
```
this method is not used

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/usecase/PushUseCase.java:
```diff
@@ -34,13 +37,19 @@
     private IPushController mPushController;
     private Callback mCallback;
+    private Credentials credentials;
+    private Context context;
```
An use case should not have android dependencies

xurxodev commented on app/src/main/java/org/eyeseetea/malariacare/domain/usecase/UpdateQuarantineSurveyStatusUseCase.java:
```diff
@@ -0,0 +1,164 @@
+import java.util.Date;
+import java.util.List;
+
+public class UpdateQuarantineSurveyStatusUseCase implements UseCase{
```
This really is not a use case, the use case is push use case. An use case should can be execute independently and this use case only is executed during PushUseCase then is not a use case

xurxodev reviewed (changes requested):
@idelcano I have requested you changes to review

idelcano committed:
Add read police strategy to get orgunits and use OrgUnitLocalDataSource

idelcano committed:
fix repositories

idelcano committed:
update filters

idelcano committed:
remove survey checker and use case

idelcano committed:
Remove AUtils check connection and use connectivity manager

idelcano committed:
add new params in factories

idelcano committed:
Added api mapper and test

idelcano committed:
refactor api mapper

idelcano committed:
fix references

idelcano committed:
refactor and add api mapper test

idelcano committed:
create api event test and refactor method

idelcano committed:
Added api events tests

idelcano committed:
Convert new test to kotlin

idelcano committed:
refactor kotlin name tests

idelcano committed:
Clean code

idelcano committed:
add kotlin support in gradle

idelcano committed:
remove unused methods

idelcano committed:
fix typo

idelcano commented:
I added the first test using kotlin, i added the library following the IDe steps, but at the end the IDE is saying the version of the kotlin library in the  light sdk should be removed.

@xurxodev can you take a look to this error with gradle, kotlin, and the new sdk?

xurxodev reviewed (changes requested):
@idelcano can you review failed Travis compilation?

idelcano committed:
Merge branch 'development' into feature-create_remote_data_source_to_call_getevents
